src/Adapter/Product/ProductDuplicator.php

```diff
@@ -0,0 +1,622 @@
+
+use Category;
+use GroupReduction;
+use Image;
+use Pack;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\CannotDuplicateProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\CannotUpdateProductException;
+use PrestaShop\PrestaShop\Core\Domain\Product\ProductSettings;
+use PrestaShop\PrestaShop\Core\Domain\Product\ValueObject\ProductId;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Multistore\MultistoreContextCheckerInterface;
+use PrestaShop\PrestaShop\Core\Util\String\StringModifierInterface;
+use PrestaShopException;
+use Product;
+use Search;
+use Shop;
+use ShopGroup;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Duplicates product
+ */
+class ProductDuplicator
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @var bool
+     */
+    private $isSearchIndexationOn;
+
+    /**
+     * @var MultistoreContextCheckerInterface
+     */
+    private $multistoreContextChecker;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var StringModifierInterface
+     */
+    private $stringModifier;
+
+    /**
+     * @param ProductRepository $productRepository
+     * @param HookDispatcherInterface $hookDispatcher
+     * @param bool $isSearchIndexationOn
+     * @param MultistoreContextCheckerInterface $multistoreContextChecker
+     * @param TranslatorInterface $translator
+     * @param StringModifierInterface $stringModifier
+     */
+    public function __construct(
+        ProductRepository $productRepository,
+        HookDispatcherInterface $hookDispatcher,
+        bool $isSearchIndexationOn,
+        MultistoreContextCheckerInterface $multistoreContextChecker,
+        TranslatorInterface $translator,
+        StringModifierInterface $stringModifier
+    ) {
+        $this->productRepository = $productRepository;
+        $this->hookDispatcher = $hookDispatcher;
+        $this->isSearchIndexationOn = $isSearchIndexationOn;
+        $this->multistoreContextChecker = $multistoreContextChecker;
+        $this->translator = $translator;
+        $this->stringModifier = $stringModifier;
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return ProductId new product id
+     *
+     * @throws CannotDuplicateProductException
+     * @throws CannotUpdateProductException
+     * @throws CoreException
+     */
+    public function duplicate(ProductId $productId): ProductId
+    {
+        //@todo: don't forget to add hooks from PrestaShopBundle\Controller\Admin\ProductController L1063 when new controller is implemented
```

Comment:
hmm ok makes sense.
btw I've noticed that `...duplicateAfter` hook only passes the old product id, but it would also make sense to pass the new (recently created duplicate) product id too.
do you think we could introduce it? 
['product_id => $id, 'new_product_id' => $duplicateId]

Reply:
Yes, good idea